libraries/ms-common/src/stm32f0xx/soft_timer.c

```diff
@@ -0,0 +1,211 @@
+#include "soft_timer.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "critical_section.h"
+#include "interrupt.h"
+#include "status.h"
+#include "stm32f0xx_interrupt.h"
+#include "stm32f0xx_rcc.h"
+#include "stm32f0xx_tim.h"
+
+typedef struct Timer {
+  uint32_t duration;
+  void *context;
+  bool inuse;
+  SoftTimerCallback callback;
+} Timer;
+
+static volatile Timer s_soft_timer_array[SOFT_TIMER_MAX_TIMERS];
+static volatile SoftTimerID s_active_timer_id;
+static volatile uint32_t s_last_update_count;
+static volatile uint32_t s_active_timer_duration;
+
+static void prv_start_timer(uint32_t duration_us) {
+  // Clear any pending interrupts, this will always occur in an interrupt or new timer getting
+  // started in a critical section so we need to prevent a accidental trigger if the time expires in
+  // the middle of the critical section and we started a new timer which falls before it.
+  TIM_ClearITPendingBit(TIM2, TIM_IT_CC1);
+
+  TIM_Cmd(TIM2, DISABLE);
+  uint32_t curr_time = TIM_GetCounter(TIM2);
+  if (curr_time + duration_us > UINT32_MAX) {
+    TIM_SetCompare1(TIM2, curr_time + duration_us - UINT32_MAX);
+  } else {
+    TIM_SetCompare1(TIM2, curr_time + duration_us);
+  }
+  s_active_timer_duration = duration_us;
+  TIM_Cmd(TIM2, ENABLE);
+}
+
+static uint32_t prv_soft_timer_update_period(void) {
+  uint32_t counter = TIM_GetCounter(TIM2);
+  if (counter - s_last_update_count > 0) {
```

Comment:
uint32_t - uint32_t will always be >= 0, so you're only checking if they're the same.